File: python/9_14_2022/binary_search.py
from random import randrange, sample


def binary_search(arr, key):
    low = 0
    high = len(arr)
    while low <= high:
        mid = (low + high) // 2
        if arr[mid] > key:
            high = mid - 1
        elif arr[mid] < key:
            low = mid + 1
        else:
            return mid
    return -1


def test_case():  # TODO: Use pytest to run tests, unit_test package
    arr = [1141, 4222, 4320, 4971, 5627, 7428, 7978, 8152, 8265, 9358]
    key = 7
    assert key == binary_search(arr, arr[key])


if __name__ == '__main__':
    arr = [1, 2, 3, 4, 5]
    arr = sample(range(0, 10000), 100)
    arr.sort()  # TODO: Replace with quicksort or heapsort, timeit
    key_index = randrange(0, 100)
    # print(arr)
    res = binary_search(arr, arr[key_index])
    print(arr, res, key_index)
    assert key_index == res
    print(res)
    # test_case()

File: java/d9162022/BinarySearch.java
package d9162022;

import java.util.Arrays;

public class BinarySearch {
    public int do_search(int []arr, int val, Boolean isSorted){
        if (!isSorted){
            Arrays.sort(arr);
        }
        int low = 0;
        int high = arr.length;
        while (low <= high){
            int mid = (low + high) / 2;
            if (arr[mid] > val){
                high = mid - 1;
            }
            else if(arr[mid] < val){
                low = mid + 1;
            }
            else{
                return mid;
            }
        }
        return -1;
    }
}
